### PR TITLE
feat(dot): add support for staking transactions

### PR DESCRIPTION
**Pull request by mcalcano-bg**
feat(sdk-coin-dot): add support for staking transactions
    
add support for batchAll

add support for staking extrinsics on batchAll transactions

add support for bondExtra

add support for removeProxy

add support for createPure (anonymous proxies were renamed
to pure proxies and the extrinsic to create them was renamed
to createPure)

add support for computing inputs and outputs for staking
transactions

add fakeSign() to Transaction to serialize a transaction
with a dummy signature. This can be used to estimate
transaction fees

update metadata rpc for Westend to the latest version

add support for addresses decoded from raw transactions as
strings or objects ({id: string}). The way addresses are decoded
from raw transactions has changed.

BREAKING CHANGE: Uddate the 'westendMetadataRpc' constant to
spec version 9320. Update PolkadotTestnet.specVersion to
9320 and PolkadotTestnet.txVersion to 14.

Ticket: BOS-469

### CODE REVIEW

File: modules/sdk-coin-dot/src/lib/batchTransactionBuilder.ts [hunk 1]
```diff
@@ -69,18 +113,33 @@ export class BatchTransactionBuilder extends TransactionBuilder {
   /** @inheritdoc */
   protected fromImplementation(rawTransaction: string): Transaction {
     const tx = super.fromImplementation(rawTransaction);
-    if (this._method?.name === MethodNames.Batch) {
+    if (this._method?.name === MethodNames.Batch || this._method?.name === MethodNames.BatchAll) {
+      if (this._method?.name === MethodNames.BatchAll) {
```
Comment (margueriteblair) [inline modules/sdk-coin-dot/src/lib/batchTransactionBuilder.ts:117]: for my own understanding, what's the difference between Batch and BatchAll here?

**Review by margueriteblair — DISMISSED** [review]
Overall, looks good to me. Tests seem comprehensive. I'd personally love to see a bit more of an explanation in the commit comment about how the proxyBuilder and stakingBuilder operate and how they're working together.

File: modules/sdk-coin-dot/src/lib/iface.ts [hunk 2]
```diff
@@ -142,12 +157,73 @@ export interface AddAnonymousProxyArgs {
  */
 export type BatchCallObject = {
   callIndex: string;
-  args: Record<string, any>;
+  args:
+    | Record<string, any>
```
Comment (byhow) [inline modules/sdk-coin-dot/src/lib/iface.ts:161]: nit: would love to see `any` go away; doesn't have to be in this PR since its not the scope

**Review by byhow — APPROVED** [review]

File: modules/sdk-coin-dot/package.json [hunk 3]
```diff
@@ -54,6 +54,7 @@
     "hi-base32": "^0.5.1",
     "joi": "^17.4.0",
     "lodash": "^4.17.15",
+    "strip-hex-prefix": "^1.0.0",
```
Comment (mmcshinsky-bitgo) [inline modules/sdk-coin-dot/package.json:57]: Can we bring this in as a utility? The function is so simple that I don't think we need an npm install to do this.

File: modules/sdk-coin-dot/src/lib/iface.ts [hunk 4]
```diff
@@ -142,12 +158,87 @@ export interface AddAnonymousProxyArgs {
...
 
+export interface AddAnonymousProxyBatchCallArgs {
+  proxy_type: ProxyType;
+  index: number;
+  delay: number;
+}
+
+export interface AddProxyBatchCallArgs {
+  delegate: string | AccountId;
+  proxy_type: ProxyType;
```
Comment (mmcshinsky-bitgo) [inline modules/sdk-coin-dot/src/lib/iface.ts:180]: Let's keep consistent with our naming in camelCase unless there is a specific reason it needs to be this weay.

File: modules/sdk-coin-dot/src/lib/iface.ts [hunk 5]
```diff
@@ -142,12 +158,87 @@ export interface AddAnonymousProxyArgs {
...
+  delay: number;
+}
+
+export type AccountId = { id: string };
+
+export function isAccountId(value: string | AccountId): value is AccountId {
```
Comment (mmcshinsky-bitgo) [inline modules/sdk-coin-dot/src/lib/iface.ts:186]: Let's move functions out of type definition files.

**Review by mmcshinsky-bitgo — CHANGES_REQUESTED** [review]

File: modules/sdk-coin-dot/src/lib/batchTransactionBuilder.ts (hunk 1 above)
Comment (mcalcano-bg) [reply modules/sdk-coin-dot/src/lib/batchTransactionBuilder.ts:119]: > [utility.batchAll()](https://github.com/paritytech/substrate/blob/ded44948e2d5a398abcb4e342b0513cb690961bb/frame/utility/src/lib.rs#L284) does not commit if one of the calls in the batch fails.
> 
> The whole transaction will rollback and fail if any of the calls failed.
> 
> [utility.batch()](https://github.com/paritytech/substrate/blob/ded44948e2d5a398abcb4e342b0513cb690961bb/frame/utility/src/lib.rs#L177) commits each successful call regardless if a call fails.
> 
> If a call failed and the batch was interrupted, then the BatchInterrupted event is deposited, along with the number of successful calls made and the error of the failed call. If all were successful, then the BatchCompleted event is deposited.

Taken from https://substrate.stackexchange.com/questions/2084/what-is-the-difference-between-the-extrinsics-utility-batch-and-utility-batcha

**Comment by mcalcano-bg** [comment]
> Overall, looks good to me. Tests seem comprehensive. I'd personally love to see a bit more of an explanation in the commit comment about how the proxyBuilder and stakingBuilder operate and how they're working together.

Technically they don't work together here at the SDK level, but the client using the SDK, e.g. Staking Service, can use the StakingBuilder to stake and then use the AddressInitializationBuilder (not the greatest name) to add a proxy
for the Controller address, this way a third party can nominate or unnominate validators on behalf of the Controller address without having access to it. Once the client unstakes we can use the RemoveProxyBuilder to remove the proxy 
that was previously setup.

File: modules/sdk-coin-dot/package.json (hunk 3 above)
Comment (mcalcano-bg) [reply modules/sdk-coin-dot/package.json:57]: Fixed.

File: modules/sdk-coin-dot/src/lib/iface.ts (hunk 4 above)
Comment (mcalcano-bg) [reply modules/sdk-coin-dot/src/lib/iface.ts:180]: This is the shape of the object used by the underlying library we are using to decode transactions.

File: modules/sdk-coin-dot/src/lib/iface.ts [hunk 6]
```diff
@@ -4,18 +4,30 @@ import { BaseTxInfo, TypeRegistry, DecodedUnsignedTx } from '@substrate/txwrappe
...
+  Proxy = 'proxy',
+  Staking = 'staking',
+}
+
 /**
  * Method names for the transaction method. Names change based on the type of transaction e.g 'bond' for the staking transaction
  */
 export enum MethodNames {
   AddProxy = 'addProxy',
+  RemoveProxy = 'removeProxy',
```
[the remark this answers is not in the document]
Comment (mcalcano-bg) [reply modules/sdk-coin-dot/src/lib/iface.ts:20]: Not sure what you mean, can you rephrase that?

File: modules/sdk-coin-dot/src/lib/iface.ts (hunk 5 above)
Comment (mcalcano-bg) [reply modules/sdk-coin-dot/src/lib/iface.ts:186]: Fixed.

File: modules/sdk-coin-dot/src/lib/utils.ts [hunk 7]
```diff
@@ -264,6 +279,72 @@ export class Utils implements BaseUtils {
     return (arg as TransferAllArgs).dest?.id !== undefined && (arg as TransferAllArgs).keepAlive !== undefined;
   }
 
+  isBatch(arg: TxMethod['args']): arg is BatchArgs {
```
[the remark this answers is not in the document]
Comment (mcalcano-bg) [reply modules/sdk-coin-dot/src/lib/utils.ts:282]: Fixed.

File: modules/statics/src/networks.ts [hunk 8]
```diff
@@ -341,10 +341,10 @@ class PolkadotTestnet extends Testnet implements DotNetwork {
   explorerUrl = 'https://westend.subscan.io/extrinsic/';
   specName = 'westend' as PolkadotSpecNameType;
   genesisHash = '0xe143f23803ac50e8f6f8e62695d1ce9e4e1d68aa36c1cd2cfd15340213f3423e';
-  specVersion = 9160;
+  specVersion = 9320;
   chainName = 'Westend';
   metadataRpc = westendMetadataRpc as `0x${string}`;
-  txVersion = 9;
+  txVersion = 14;
```
[the remark this answers is not in the document]
Comment (mcalcano-bg) [reply modules/statics/src/networks.ts:347]: Not sure, how do we handle breaking changes?

File: modules/sdk-coin-dot/src/lib/utils.ts [hunk 9]
```diff
@@ -264,6 +279,139 @@ export class Utils implements BaseUtils {
...
+   *
+   * @return true if arg is of type BatchArgs and the calls of the batch are unstaking calls: a remove
+   * proxy call (removeProxy), followed by a chill call, and an unstake call (unbond), false otherwise.
+   */
+  isUnstakingBatch(arg: TxMethod['args']): arg is BatchArgs {
+    const calls = (arg as BatchArgs).calls;
+    if (calls !== undefined) {
+      return (
+        calls.length === 3 &&
+        this.isRemoveProxyBatchCallArgs(calls[0].args) &&
```
Comment (jicli) [inline modules/sdk-coin-dot/src/lib/utils.ts:356]: what is a chill call, do we not need to verify it?

File: modules/sdk-coin-dot/src/lib/transaction.ts [hunk 10]
```diff
@@ -18,18 +19,34 @@ import { KeyPair } from './keyPair';
...
   UnstakeArgs,
   WithdrawUnstakedArgs,
 } from './iface';
+import { getAddress, getDelegateAddress } from './iface_utils';
 import utils from './utils';
+import BigNumber from 'bignumber.js';
+import { Vec } from '@polkadot/types';
+import { PalletConstantMetadataV14 } from '@polkadot/types/interfaces';
+
+const FAKE_SIGNATURE = `0x${Buffer.from(new Uint8Array(256).fill(1)).toString('hex')}`;
```
Comment (jicli) [inline modules/sdk-coin-dot/src/lib/transaction.ts:43]: nit : this doesnt have to be global variable?

File: modules/sdk-coin-dot/src/lib/iface.ts [hunk 11]
```diff
@@ -4,18 +4,30 @@ import { BaseTxInfo, TypeRegistry, DecodedUnsignedTx } from '@substrate/txwrappe
...
   Proxy = 'proxy',
   Bond = 'bond',
+  BondExtra = 'bondExtra',
   TransferAll = 'transferAll',
   TransferKeepAlive = 'transferKeepAlive',
   Unbond = 'unbond',
   Anonymous = 'anonymous',
+  PureProxy = 'createPure', // Anonymous proxies were renamed to pure proxies
```
Comment (jicli) [inline modules/sdk-coin-dot/src/lib/iface.ts:28]: should me make them the same name ?
```suggestion
  PureProxy = 'pureProxy', // Anonymous proxies were renamed to pure proxies
```

File: modules/sdk-coin-dot/src/lib/iface.ts (hunk 4 above)
Comment (lcovar) [reply modules/sdk-coin-dot/src/lib/iface.ts:180]: A comment saying that ^ in the code would be helpful.

**Review by jicli — DISMISSED** [review]
LGTM, approved, but we kinda need to identify if it's a breaking change as well

File: modules/statics/src/networks.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (mmcshinsky-bitgo) [reply modules/statics/src/networks.ts:347]: You'll need to add a BREAKING CHANGE: description to the comment footer. https://www.conventionalcommits.org/en/v1.0.0/

File: modules/sdk-coin-dot/src/lib/iface.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (mmcshinsky-bitgo) [reply modules/sdk-coin-dot/src/lib/iface.ts:20]: Is there documentation for a user to fully explain what each name does if they use it?

File: modules/sdk-coin-dot/src/lib/iface.ts (hunk 4 above)
Comment (mcalcano-bg) [reply modules/sdk-coin-dot/src/lib/iface.ts:258]: Added a comment.

File: modules/sdk-coin-dot/src/lib/iface.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (mcalcano-bg) [reply modules/sdk-coin-dot/src/lib/iface.ts:30]: Added comments and links to the docs.

File: modules/statics/src/networks.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (mcalcano-bg) [reply modules/statics/src/networks.ts:347]: Fixed.

File: modules/sdk-coin-dot/src/lib/utils.ts (hunk 9 above)
Comment (mcalcano-bg) [reply modules/sdk-coin-dot/src/lib/utils.ts:357]: Good point, chill has no arguments so it's an empty, added a check for it.

File: modules/sdk-coin-dot/src/lib/transaction.ts (hunk 10 above)
Comment (mcalcano-bg) [reply modules/sdk-coin-dot/src/lib/transaction.ts]: Fixed.

File: modules/statics/src/networks.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (mmcshinsky-bitgo) [reply modules/statics/src/networks.ts:347]: Remove the space between BREAKING CHANGE and the ticket number to put it into the footer:

```
BREAKING CHANGE: Uddate the 'westendMetadataRpc' constant to
spec version 9320. Update PolkadotTestnet.specVersion to
9320 and PolkadotTestnet.txVersion to 14.
Ticket: BOS-469
```

[the remark this answers is not in the document]
Comment (mcalcano-bg) [reply modules/statics/src/networks.ts:347]: Fixed.

**Review by jicli — APPROVED** [review]
